viewing_party/party.py

```diff
@@ -1,23 +1,223 @@
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    users_data = user_data
+    my_movies = users_data["watched"]
+    friends_watched_movies = friends_watched(user_data)
+    unique_list = []   
+    
+
+    if my_movies:
+        for movie in my_movies:
+            if movie not in friends_watched_movies:
+                unique_list.append(movie)
+        return unique_list
+    else:
+        return unique_list  
+
+def get_friends_unique_watched(user_data):
+
+    users_data = user_data
+    your_watched_movies = users_data["watched"]
+    friends_watched_movies = friends_watched(user_data)
+    friends_unique_watched = []
+
+    for movie in friends_watched_movies:
+        if movie not in your_watched_movies:
+            friends_unique_watched.append(movie)
+    return friends_unique_watched
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    users_data = user_data
+    recommended_movies = []
+    friends_unique_watched = get_friends_unique_watched(user_data)
+
+    for movie in friends_unique_watched: 
+        if movie["host"] in users_data["subscriptions"]:
+            recommended_movies.append(movie)
+    return recommended_movies
+
+
+def get_new_rec_by_genre(user_data):
+
+    users_data = user_data
+    new_rec_by_genre = []
+    most_watched_genre = get_most_watched_genre(users_data)
+    friends_unique_watched = get_friends_unique_watched(users_data)
+
+    for movie in friends_unique_watched:
+        if movie["genre"] == most_watched_genre:
+            new_rec_by_genre.append(movie)
+
+    return new_rec_by_genre
+
+
+def get_rec_from_favorites(user_data):
+
+    users_data = user_data
+    favorites = users_data["favorites"]
+    recommend = []
+    friends_watched_movies = friends_watched(users_data)
+
+    for movie in favorites:
+        if movie not in friends_watched_movies:
+            recommend.append(movie)
+    return recommend
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+
+    users_data = user_data
+    new_rec_by_genre = []
+    most_watched_genre = get_most_watched_genre(users_data)
+    friends_unique_watched = get_friends_unique_watched(users_data)
+
+    for movie in friends_unique_watched:
+        if movie["genre"] == most_watched_genre:
+            new_rec_by_genre.append(movie)
+    return new_rec_by_genre
+
+
+def get_rec_from_favorites(user_data):
+
+    users_data = user_data
+    favorites = users_data["favorites"]
+    recommend = []
+    friends_watched_movies = friends_watched(users_data)
+
+    for movie in favorites:
+        if movie not in friends_watched_movies:
+            recommend.append(movie)
+    return recommend

```

Comment:
This approach starts from the favorite movies list, and only recommends it if it's not in the list of movies watched by the friends. An alternative approach would be to start from the unique user's watched movies, and only recommend a movie it it appears in the user's favorite list. And for either approach, we could use title sets to improve the performance of lookups (compared to `in` on a list).